lib/mobile_app_backend/global_data_cache.ex

```diff
@@ -24,40 +26,132 @@ defmodule MobileAppBackend.GlobalDataCache do
+
+  def start_link(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).start_link(opts)
+  end
+
+  def init(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).init(opts)
+  end
+
+  def default_key do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).default_key()
+  end
+
+  def get_data(key \\ default_key()) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).get_data(key)
+  end
+
+  def route_ids_for_stops(stop_ids, key \\ default_key()) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).route_ids_for_stops(stop_ids, key)
+  end
+end
+
+defmodule MobileAppBackend.GlobalDataCache.Impl do
+  use GenServer
+  alias MBTAV3API.{JsonApi, Repository}
+  alias MobileAppBackend.GlobalDataCache
+  alias MobileAppBackend.GlobalDataCache.State
+
+  @behaviour GlobalDataCache
+
+  @impl true
+  def default_key, do: MobileAppBackend.GlobalDataCache
+
+  @spec start_link(keyword()) :: :ignore | {:error, any()} | {:ok, pid()}
+  def start_link(opts \\ []) do
     opts = Keyword.merge([key: default_key()], opts)
-    GenServer.start_link(__MODULE__, opts)
+    GenServer.start_link(GlobalDataCache, opts)
   end
 
-  @spec get_data(key()) :: data()
+  @impl true
   def get_data(key \\ default_key()) do
     :persistent_term.get(key, nil) || update_data(key)
   end
 
+  @impl true
+  def route_ids_for_stops(stop_ids, key \\ default_key()) do
+    data = :persistent_term.get(key, nil) || update_data(key)
+
+    if is_nil(data) do
+      :error
+    else
+      %{pattern_ids_by_stop: pattern_ids_by_stop, route_patterns: route_patterns} =
+        data
+
+      route_pattern_ids =
+        pattern_ids_by_stop
+        |> Map.take(stop_ids)
+        |> Enum.flat_map(fn {_stop_id, pattern_ids} -> pattern_ids end)
+        |> Enum.uniq()
+
+      routes =
+        route_patterns
+        |> Map.take(route_pattern_ids)
+        |> Enum.map(fn {_route_pattern_id, pattern} -> pattern.route_id end)
+        |> Enum.uniq()
+
+      routes
+    end
+  end
+
   @impl GenServer
-  def init(opts) do
-    opts = Keyword.merge(Application.get_env(:mobile_app_backend, __MODULE__), opts)
+  def init(opts \\ []) do
+    opts = Keyword.merge(Application.get_env(:mobile_app_backend, GlobalDataCache), opts)
 
     state = %State{
       key: opts[:key],
       update_ms: opts[:update_ms] || :timer.minutes(5)
     }
 
+    Process.send_after(self(), :recalculate, :timer.seconds(1))
```

Comment:
It's weird that this wasn't here before. This is really tough to test locally, since you can only know it's working if GTFS actually changes, but that might mean this never actually worked and was only ever calculating the global data once. Oops!

Reply:
Added some light tests for this by checking that the message is sent